app/lib/aether/aether.coffee

```diff
@@ -0,0 +1,244 @@
+    @reset()
+    return language
+
+  # Resets the state of Aether, readying it for a fresh transpile.
+  reset: ->
+    @problems = errors: [], warnings: [], infos: []
+    @style = {}
+    @flow = {}
+    @metrics = {}
+    @pure = null
+
+  # Convert to JSON so we can pass it across web workers and HTTP requests and store it in databases and such.
+  serialize: ->
+    _.pick @, ['originalOptions', 'raw', 'pure', 'problems', 'flow', 'metrics', 'style', 'ast']
+
+  # Convert a serialized Aether instance back from JSON.
+  @deserialize: (serialized) ->
+    aether = new Aether serialized.originalOptions
+    aether[prop] = val for prop, val of serialized when prop isnt "originalOptions"
+    aether
+
+  # Performs quick heuristics to determine whether the code will run or produce compilation errors.
+  # If thorough, it will perform detailed linting and return false if there are any lint errors.
+  canTranspile: (rawCode, thorough=false) ->
+    return true if not rawCode # blank code should compile, but bypass the other steps
+    return false if @language.obviouslyCannotTranspile rawCode
+    return true unless thorough
+    @lint(rawCode, @).errors.length is 0
+
+  # Determine whether two strings of code are significantly different.
+  # If careAboutLineNumbers, we strip trailing comments and whitespace and compare line count.
+  # If careAboutLint, we also lint and make sure lint problems are the same.
+  hasChangedSignificantly: (a, b, careAboutLineNumbers=false, careAboutLint=false) ->
+    return true unless a? and b?
+    return false if a is b
+    return true if careAboutLineNumbers and @language.hasChangedLineNumbers a, b
+    return true if careAboutLint and @hasChangedLintProblems a, b
+    # If the simple tests fail, we compare abstract syntax trees for equality.
+    @language.hasChangedASTs a, b
+
+  # Determine whether two strings of code produce different lint problems.
+  hasChangedLintProblems: (a, b) ->
+    aLintProblems = ([p.id, p.message, p.hint] for p in @getAllProblems @lint a)
+    bLintProblems = ([p.id, p.message, p.hint] for p in @getAllProblems @lint b)
+    return not _.isEqual aLintProblems, bLintProblems
+
+  # Return a beautified representation of the code (cleaning up indentation, etc.)
+  beautify: (rawCode) ->
+    @language.beautify rawCode, @
+
+  # Transpile it. Even if it can't transpile, it will give syntax errors and warnings and such. Clears any old state.
+  transpile: (@raw) ->
+    @reset()
+    rawCode = @raw
+    @problems = @lint rawCode
+    @pure = @purifyCode rawCode
+    @pure
+
+  # Perform some fast static analysis (without transpiling) and find any lint problems.
+  lint: (rawCode) ->
+    lintProblems = errors: [], warnings: [], infos: []
+    @addProblem problem, lintProblems for problem in @language.lint rawCode, @
+    lintProblems
+
+  # Return a ready-to-interpret function from the parsed code.
+  createFunction: ->
```

Comment:
This is a method that is called from the Level Editor a lot. I don't understand it fully yet. However it appears to use esper within it.